qcodes_contrib_drivers/drivers/Thorlabs/APT.py

```diff
@@ -35,73 +47,342 @@ def __init__(self, dll_path: Optional[str] = None, verbose: bool = False, event_
         self.apt_init()
         self.enable_event_dlg(event_dialog)
 
-    def error_check(self, code, function_name=''):
+    def error_check(self, code: int, function_name: str = "") -> None:
+        """Analyzes a functions return code to check, if the function call to APT.dll was
+        successful. If an error occurred, this function throws an ThorlabsException.
+
+        Args:
+            code: The called function's return code
+            function_name: Name of the called function
+
+        Throws:
+            ThorlabsException: Thrown, if the return code indicates that an error has occurred.
+        """
         if code == self._success_code:
             if self.verbose:
-                print("APT: [%s]: %s" % (function_name, 'OK - no error'))
+                print("APT: [{}]: {}".format(function_name, "OK - no error"))
         else:
-            print("APT: [%s]: Unknown code: %s" % (function_name, code))
-            raise Exception()
+            raise ThorlabsException("APT: [{}]: Unknown code: {}".format(function_name, code))
 
-    def apt_clean_up(self):
+    def apt_clean_up(self) -> None:
+        """Cleans up the resources of APT.dll"""
         code = self.dll.APTCleanUp()
         self.error_check(code, 'APTCleanUp')
 
-    def apt_init(self):
+    def apt_init(self) -> None:
+        """Initialization of APT.dll"""
         code = self.dll.APTInit()
         self.error_check(code, 'APTInit')
 
-    def enable_event_dlg(self, enable):
+    def list_available_devices(self, hw_type: Union[int, ThorlabsHWType] = None) \
+            -> List[Tuple[int, int, int]]:
+        """Lists all available Thorlabs devices, that can connect to the APT server.
+
+        Args:
+            hw_type: If this parameter is passed, the function only searches for a certain device
+                     model. Otherwise (if the parameter is None), it searches for all Thorlabs
+                     devices.
+
+        Returns:
+            A list of tuples. Each list-element is a tuple of 3 ints, containing the device's
+            hardware type, device id and serial number: [(hw type id, device id, serial), ...]
+        """
+        devices = []
+        count = ctypes.c_long()
+
+        if hw_type is not None:
+            if isinstance(hw_type, ThorlabsHWType):
+                hw_type_range = [hw_type.value]
+            else:
+                hw_type_range = [int(hw_type)]
+        else:
+            hw_type_range = list(range(100))
+
+        for hw_type_id in hw_type_range:
```

Comment:
Is there any way that we can reduce this logic flow and nested loops?

Reply:
Unfortunately not, or at least I have no idea how, but I will add some comments to make it clearer.

Reply:
Ok thanks. :)